### PR TITLE
fix(css): clarify counter-set vs counter-reset

### LINKED ISSUE #13293: Issue with "counter-reset": What is the difference in behavior between the latest Firefox (82 or later) and the older 81 or earlier or other browsers?
MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/counter-reset

Maybe the page to fix is this: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Counter_Styles/Using_CSS_counters

#### What information was incorrect, unhelpful, or incomplete?

- (p1) I know that starting with Firefox 82, the behavior of counter-reset has changed from 81 and earlier versions.
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1754521
- (p2) For example, the results of the CSS Demo for counter-reset are different between Firefox (the latest version) and Google Chrome.
    - https://developer.mozilla.org/en-US/docs/Web/CSS/counter-reset
- (p3) The ol and ul elements use CSS counter(list-item) to calculate the numbering in the marker.
    - https://drafts.csswg.org/css-lists/#propdef-list-style-type
- (p4) I'd like to see a clear sample of the difference in behavior between counter-reset and counter-set.
- (p5) In addition, I would like to see a more helpful example of how to use reversed().

It is my hope that the documentation will clearly explain why each browser behaves the way it does.

In particular, I think the following current statement is wrong.

> The value to reset the counter to on each occurrence of the element. Defaults to 0 if not specified. 

The latest version of Firefox doesn't seem to reset to on each occurrence of the element. This may just be because it resets the counter, but creates another instance, which makes it look like the counter is not being reset.

#### Related specs

- https://www.w3.org/TR/css-lists-3/#counter-reset
- https://drafts.csswg.org/css-lists/#counter-reset
- https://github.com/w3c/csswg-drafts/issues/4244
- https://github.com/w3c/csswg-drafts/issues/5477
- https://html.spec.whatwg.org/#lists
- https://drafts.csswg.org/css-lists/#propdef-list-style-type

#### Related links

- https://bugzilla.mozilla.org/show_bug.cgi?id=1548753
- https://github.com/mdn/content/issues/3737
- https://github.com/mdn/content/issues/6360
- https://bugzilla.mozilla.org/show_bug.cgi?id=1754521
- https://bugzilla.mozilla.org/show_bug.cgi?id=1757269

#### What did you expect to see?

I don't know what the specs say about subtle nuances and what is the right way to behave.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/css/counter-reset`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/counter-reset
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/counter-reset/index.md
* Last commit: https://github.com/mdn/content/commit/c2349cdb4688f6e5a8c90ae12addada359dbca4c
* Document last modified: 2022-01-28T02:17:46.000Z

</details>

### PR DESCRIPTION
**Pull request by OnkarRuikar**
- fix https://github.com/mdn/content/issues/13293

The PR:
- Highlights the difference between `counter-set` and `counter-reset` properties using an example.
- Corrects the wrong wording for both properties in various places.
- Adds example for `reversed()` counter.

### CODE REVIEW

File: files/en-us/web/css/counter-reset/index.md [hunk 1]
```diff
@@ -134,6 +135,47 @@ ol {
...
+
+In the following example, we've declared a reversed counter named 'priority'. The counter is being used to number five tasks.
+
+```html
+<ul class="stack">
+  <li>Task one</li>
+  <li>Task two</li>
+  <li>Task three</li>
+  <li>Task four</li>
+  <li>Task five</li>
```
Comment (bsmth) [inline files/en-us/web/css/counter-reset/index.md:148]: One thing I've found convenient in counting and numbering examples is to use Alpha for non-modified things, it helps avoid "one 1, two 3, five 6" type output which is hard to follow, sometimes:

```suggestion
  <li>Task A</li>
  <li>Task B</li>
  <li>Task C</li>
  <li>Task D</li>
  <li>Task E</li>
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 2]
```diff
@@ -129,6 +143,68 @@ The counter value is decreased by specifying a negative value for {{cssxref("cou
...
 
+### Difference between counter-set and counter-reset
+
+The {{cssxref("counter-set")}} property updates an existing counter and if no counter with the name exists then a new counter is instantiated. The {{cssxref("counter-increment")}} property _always_ creates a new counter.
+
+In the following example, we have two sub-lists inside a parent list. Each list item has been numbered using a counter named 'item'. The first sub-list uses {{cssxref("counter-set")}} property and the second sub-list uses {{cssxref("counter-reset")}} property to change the 'item' counter.
+
+```html
+<ul class="parent">
+  <li>one</li>
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:154]: My comment above is more relevant here where you get this on the top-level

```
1 one
2 two
3 three
13 four
14 five
15 six
16 seven
```

Previously I've seen this kind of thing help:

```
1 A
2 B
3 C
13 D
14 E
15 F
16 G
```

File: files/en-us/web/css/counter-reset/index.md [hunk 3]
```diff
@@ -57,7 +57,8 @@ The `counter-reset` property accepts a list of one or more space-separated count
 
 The `counter-reset` property can create both regular and, in browsers that support it, reversed counters. You can create multiple regular and reversed counters, each separated by a space. Counters can be a standalone name or a space-separated name-value pair.
 
-After creating a counter using `counter-reset`, you can adjust its value by using the {{cssxref("counter-set")}} property. This is counterintuitive because, despite its name, the `counter-reset` property is used for creating and initializing counters, while the `counter-set` property is used for resetting the value of an existing counter.
+> [!WARNING]
+> There is [a difference between `counter-reset` and `counter-set` properties](/en-US/docs/Web/CSS/CSS_counter_styles/Using_CSS_counters#difference_between_counter-set_and_counter-reset). After creating a counter using `counter-reset`, you can adjust its value by using the {{cssxref("counter-set")}} property. This is counterintuitive because, despite its name, the `counter-reset` property is used for creating and initializing counters, while the `counter-set` property is used for resetting the value of an existing counter.
```
Comment (bsmth) [inline files/en-us/web/css/counter-reset/index.md:61]: I'm not sure about this warning. You can reset a counter using:

```css
h1::before {
    content: "Chapter " counter(chapter) ". ";
    counter-increment: chapter;  /* Add 1 to chapter */
    counter-reset: section;      /* Set section to 0 */
}
```

See https://drafts.csswg.org/css-lists/#example-838dca3e

So it might be better to say `counter-set` is good for directly setting values (like skipping increments), and counter reset is for initializing and resetting.

I would keep "After creating a counter using `counter-reset`, you can adjust its value by using the {{cssxref("counter-set")}} property." but I wouldn't mention the others, especially "`counter-set` property is used for resetting".

File: files/en-us/web/css/counter-set/index.md [hunk 4]
```diff
@@ -7,9 +7,11 @@ browser-compat: css.properties.counter-set
 
 {{CSSRef}}
 
-The **`counter-set`** [CSS](/en-US/docs/Web/CSS) property sets [CSS counters](/en-US/docs/Web/CSS/CSS_counter_styles/Using_CSS_counters) to the given values.
+The **`counter-set`** [CSS](/en-US/docs/Web/CSS) property sets existing [CSS counters](/en-US/docs/Web/CSS/CSS_counter_styles/Using_CSS_counters) on the element to the given values.
```
Comment (bsmth) [inline files/en-us/web/css/counter-set/index.md:10]: Maybe we can leave out "existing":

>They only [instantiate](https://drafts.csswg.org/css-lists/#instantiate-counter) new counters if there is no counter of the given name on the element yet.

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 5]
```diff
@@ -8,15 +8,21 @@ spec-urls: https://drafts.csswg.org/css-lists/#auto-numbering
 {{CSSRef}}
 
 **CSS counters** let you adjust the appearance of content based on its location in a document.
-For example, you can use counters to automatically number the headings in a webpage, or to change the numbering on ordered lists.
+For example, you can use counters to automatically number the headings on a webpage or to change the numbering on ordered lists.
+
+Counters are, in essence, variables maintained by CSS whose values may be incremented or decremented by CSS rules that track how many times they're used. The following things affect the counter values on an element:
+
+1. Counters are inherited from the parent element or received from a previous sibling.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:15]: With inheritance, it might be good to elaborate what that means.

https://drafts.csswg.org/css-lists/#inheriting-counters

Maybe an example is worth it (as a follow-up?)

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 6]
```diff
@@ -8,15 +8,21 @@ spec-urls: https://drafts.csswg.org/css-lists/#auto-numbering
...
+2. New counters are instantiated using {{cssxref("counter-reset")}} property.
+3. Counters are incremented using {{cssxref("counter-increment")}} property.
+4. Counter values are changed using {{cssxref("counter-set")}} property.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:18]:
```suggestion
4. Counters are directly set to a value using the {{cssxref("counter-set")}} property.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 7]
```diff
@@ -8,15 +8,21 @@ spec-urls: https://drafts.csswg.org/css-lists/#auto-numbering
...
 
-Counters are, in essence, variables maintained by CSS whose values may be incremented or decremented by CSS rules that track how many times they're used.
 You can define your own named counters, and you can also manipulate the `list-item` counter that is created by default for all ordered lists.
 
 ## Using counters
 
 To use a counter it must first be initialized to a value with the {{cssxref("counter-reset")}} property.
-The counter's value can then be increased or decreased using {{cssxref("counter-increment")}} property.
+The counter's value can then be increased or decreased using the {{cssxref("counter-increment")}} property. And the counter value can also be changed using the {{cssxref("counter-set")}} property.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:25]:
```suggestion
The counter's value can be increased or decreased using the {{cssxref("counter-increment")}} property and can be directly set to a specific value using the {{cssxref("counter-set")}} property.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 8]
```diff
@@ -51,7 +57,15 @@ h3::before {
 }
 ```
 
-You can specify the value to increment or decrement the counter after the counter name, using a positive or negative number.
+You can specify the increment or decrement amount after the counter name. It can be a positive or negative number.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:60]:
```suggestion
You can specify the increment or decrement amount after the counter name. It can be a positive or negative number, but defaults to `1` if no integer is provided.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 9]
```diff
@@ -51,7 +57,15 @@ h3::before {
...
+
+Apart from increment or decrement, counters can also be explicitly set to a value using {{cssxref("counter-increment")}} property.
+
+```css
+.done::before {
+  counter-set: section;
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:66]: Want to set it to a value here?

```suggestion
  counter-set: section 20;
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 10]
```diff
@@ -129,6 +143,68 @@ The counter value is decreased by specifying a negative value for {{cssxref("cou
 > You can also use {{cssxref("counter-increment")}} to decrement a non-reversed counter.
 > The main benefit of using a reversed counter is the default initial value, and that the `list-item` counter automatically decrements reversed counters.
 
+### Difference between counter-set and counter-reset
+
+The {{cssxref("counter-set")}} property updates an existing counter and if no counter with the name exists then a new counter is instantiated. The {{cssxref("counter-increment")}} property _always_ creates a new counter.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:148]: `counter-increment` is the same as `counter-set` in terms of "if none exists, instantiate one"

```suggestion
The {{cssxref("counter-set")}} property updates an existing counter and if no counter with the name exists then a new counter is instantiated.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 11]
```diff
@@ -129,6 +143,68 @@ The counter value is decreased by specifying a negative value for {{cssxref("cou
...
+
+/* increment the counter on each list item */
+li {
+  counter-increment: item;
+}
+
+/* create a new counter for the first time */
+.parent {
+  counter-reset: item 0;
+}
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:191]: Same as:

```suggestion
/* Reset the 'item' counter */
.parent {
  counter-reset: item;
}
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md [hunk 12]
```diff
@@ -129,6 +143,68 @@ The counter value is decreased by specifying a negative value for {{cssxref("cou
...
+
+/* change the counter value */
+.sub-list-two {
+  counter-reset: item 0;
+}
+```
+
+{{EmbedLiveSample("Difference between counter-set and counter-reset", "100%", 300)}}
+
+Notice how the first sub-list items start receiving numbers starting from `11`, and the numbering is continued in the parent list. This is because the `counter-set` property updates the same 'item' counter. Then notice how the second sub-list items receive new numbering starting from '1' and the parent list items after it don't carry forward the numbering. This is because the `counter-reset` property created a new counter with the same name so the parent list items kept using the old counter.
```
Comment (bsmth) [inline files/en-us/web/css/css_counter_styles/using_css_counters/index.md:206]: I'm not sure I follow this explanation, specifically "creating a new counter of the same name". Have you had a look at this one: https://drafts.csswg.org/css-lists/#counter-nesting-example

File: files/en-us/web/css/counter-reset/index.md (hunk 3 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/counter-reset/index.md:61]: I think the spec example [has a typo](https://github.com/w3c/csswg-drafts/pull/10669), and it should use `counter-set` instead. 

> counter reset is for initializing and resetting.

The spec [doesn't say](https://drafts.csswg.org/css-lists/#valdef-counter-reset-counter-name-integer) `counter-reset` updates the existing counter if a counter with the same name exists. So `counter-reset` *can't actually* reset the same counter.

File: files/en-us/web/css/counter-set/index.md (hunk 4 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/counter-set/index.md:10]:
```suggestion
The **`counter-set`** [CSS](/en-US/docs/Web/CSS) property sets [CSS counters](/en-US/docs/Web/CSS/CSS_counter_styles/Using_CSS_counters) on the element to the given values.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md (hunk 10 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/css_counter_styles/using_css_counters/index.md:148]:
```suggestion
The {{cssxref("counter-set")}} property updates an existing counter and if no counter with the name exists then a new counter is instantiated. The {{cssxref("counter-reset")}} property _always_ creates a new counter.
```

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md (hunk 11 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/css_counter_styles/using_css_counters/index.md:191]: How can a counter be reset without creating it first? The parent is the first element that creates a counter. Perhaps we need to put the rule first in the style.

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md (hunk 5 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/css_counter_styles/using_css_counters/index.md:15]: Created a new section.

File: files/en-us/web/css/css_counter_styles/using_css_counters/index.md (hunk 12 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/css_counter_styles/using_css_counters/index.md:206]: :confused: Yes, the spec section does say that, ` instantiating a new counter on an element, which inherited an identically-named counter from its parent, creates a new counter of the same name, nested inside the existing counter. `

Comment (bsmth) [reply files/en-us/web/css/css_counter_styles/using_css_counters/index.md]: Okay I re-read and I understand now, sorry for picking on that

File: files/en-us/web/css/counter-reset/index.md (hunk 3 above)
Comment (bsmth) [reply files/en-us/web/css/counter-reset/index.md:61]: right, it's in https://drafts.csswg.org/css-lists/#increment-set 

>Manipulating Counter Values: the [counter-increment](https://drafts.csswg.org/css-lists/#propdef-counter-increment) and [counter-set](https://drafts.csswg.org/css-lists/#propdef-counter-set) properties

dear oh dear 😅

**Review by bsmth — APPROVED** [review]
Thanks Onkar, I think we can land the changes 👍🏻
